src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -3227,25 +3243,279 @@ IEnumerator IEnumerable.GetEnumerator()
             return GetEnumerator();
         }
     }
+    
+    internal interface IArrayProvider<TElement>
+    {
+        TElement[] ToArray();
+    }
+
+    internal interface IPartition<TElement> : IEnumerable<TElement>, IArrayProvider<TElement>
+    {
+        IPartition<TElement> Skip(int count);
+
+        IPartition<TElement> Take(int count);
+
+        TElement ElementAt(int index);
+
+        TElement ElementAtOrDefault(int index);
+        
+        TElement First();
+        
+        TElement FirstOrDefault();
+        
+        TElement Last();
+
+        TElement LastOrDefault();
+    }
+    
+    internal sealed class EmptyPartition<TElement> : IPartition<TElement>, IEnumerator<TElement>
```

Comment:
Using this means that the empty-by-definition cases where e.g. a start index is greater than a end index can be considered more simply. It should also be pretty efficient in itself, but that isn't as big a deal since it should also be a relatively rare case.